stdlib/public/core/UnsafeRawBufferPointer.swift.gyb

```diff
@@ -676,6 +675,44 @@ extension Unsafe${Mutable}RawBufferPointer {
     let buffer = UnsafeMutableBufferPointer(start: start, count: written)
     return (unwritten, buffer)
   }
+
+  /// Initializes the buffer's memory with the given elements, binding the
+  /// initialized memory to the elements' type.
+  ///
+  /// When calling the `initializeMemory(as:from:)` method on a buffer `b`,
+  /// the memory referenced by `b` must be uninitialized or initialized to a
+  /// trivial type, and must be properly aligned for accessing `C.Element`.
+  /// The buffer must contain sufficient memory to accommodate
+  /// `source.count`.
+  ///
+  /// This method initializes the buffer with elements from `source` until
+  /// `source` is exhausted. After calling `initializeMemory(as:from:)`,
+  /// the memory referenced by the returned `UnsafeMutableBufferPointer`
+  /// instance is bound and initialized to type `C.Element`.
+  ///
+  /// - Parameters:
+  ///   - type: The type of the elements to bind the buffer's memory to.
+  ///   - source: A collection of elements with which to initialize the buffer.
+  /// - Returns: An iterator to any elements of `source` that didn't fit in the
+  ///   buffer, and a typed buffer of the written elements. The returned
+  ///   buffer references memory starting at the same base address as this
+  ///   buffer.
+  @_alwaysEmitIntoClient
+  public func initializeMemory<C: Collection>(
+    as type: C.Element.Type, from source: C
+  ) -> (unwritten: C.Iterator, initialized: UnsafeMutableBufferPointer<C.Element>) {
+
+    let available = count / MemoryLayout<C.Element>.stride
+    _debugPrecondition(source.count <= available,
+      "insufficient space to accommodate source.count elements")
+
+    let finalCount = Swift.min(source.count, available)
```

Comment:
What if we fully embraced trapping here and made this a full `_precondition`?

Reply:
If we're calling `source.count` anyway, then this might as well be a full precondition. That causes the overload's behavior to deviate, but only in release builds and doesn't change semantics.

If you're calling `min(source.count` just for `bindMemory`, then don't bother. Overestimating there, when the input already violates a (debug) precondition will not affect memory safety in practice. A sanitizer tool could report an error, which would be fine.

I thought you were calling `min(source.count` to prevent `_copyContents` from trashing memory, which seems worthwhile even when a  debug precondition is violated.